### ISSUE TITLE
[tracking] adopt `BlockNumberProvider` for the pallets migrating to AH 

### ISSUE THREAD

**Issue by muharem** [paritytech/polkadot-sdk#6297]
Some of the pallets migrating from the Relay Chain to the Asset Hub require a more deterministic clock than what a parachain can provide. To address this, the decision was made to adapt these pallets to use a configurable block number provider, which can be set to the Relay Chain’s block number provider for greater determinism. (For more details, see the discussions at https://github.com/paritytech/polkadot-sdk/issues/3268). This approach will also simplify the migration of pallets' state from the Relay Chain to the Asset Hub since, in most cases, block numbers won’t require mapping.

### Task:

Adapt `BlockNumberProvider` for the pallets migrating to the Asset Hub:
- Introduce a `BlockNumberProvider` configuration parameter and use its `current_block_number` function as the time reference instead of `System::<T>::block_number()`.
- Ensure pallet hooks like `on_initialize(n)` are compatible with the `BlockNumberProvider` and can handle cases when the hook isn’t called on every tick (e.g., adapt conditions that rely on exact matches like `compilation_time == BlockNumberProvider::current_block_number()`).
- Review all instances of generic block number types, ensuring they reference the correct type (e.g., `BlockNumberProvider::BlockNumber` when migrated to BlockNumberProvider).
- Update documentation for calls with block number parameters, configuration settings like `BountyUpdatePeriod`, and storage fields. Documentation should clarify for users that block numbers refer to those from `Config::BlockNumberProvider`, which may come from either the local parachain or the Relay Chain.

Reference PR: https://github.com/paritytech/polkadot-sdk/pull/3970

Pallet list: 

(when you start to work on a pallet, pul your name next to it or leave a comment)

- pallet_scheduler 
- pallet_treasury - https://github.com/paritytech/polkadot-sdk/pull/3970
- pallet_child_bounties
- pallet_bounties
- pallet_referenda
- pallet_conviction_voting
- pallet_election_provider_multi_phase
- pallet_nis
- pallet_nomination_pools
- pallet_recovery
- pallet_society

**Comment by AurevoirXavier** [comment]
Love to take on some tasks once the #3970 is merged as a standard to follow.

**Comment by muharem** [comment]
@AurevoirXavier it is merged

**Comment by AurevoirXavier** [comment]
- #6446

**Comment by Doordashcon** [comment]
is anyone working `pallet-nis` atm?

**Comment by gui1117** [comment]
pallet nis make use of `total_issuance`. So migrating to AH also mean that DOT total issuance is recorded on AH. Also I don't how much we teleport asset between trusted chains, but teleporting asset does change the accuracy of the value no?

**Comment by Doordashcon** [comment]
is anyone working on `pallet-society` atm?

**Comment by PolkadotDom** [comment]
Not on the list @muharem but pallet-core-fellowship #6978

**Comment by seadanda** [comment]
> pallet nis make use of `total_issuance`. So migrating to AH also mean that DOT total issuance is recorded on AH. Also I don't how much we teleport asset between trusted chains, but teleporting asset does change the accuracy of the value no?

Teleported assets are tracked in the `CheckingAccount`, which is currently on the Relay tracking DOT which has been teleported to other chains, but will be on AH post-migration. Since pallet-nis is also migrated it shouldn't be a problem.


> is anyone working on `pallet-society` atm?

I don't think anybody is, feel free to take it on if you're interested, thanks!


> Not on the list @muharem but pallet-core-fellowship #6978 and pallet-salary #7000

This issue is focusing on the pallets being migrated from the Relay Chain to Asset Hub in the coming migration, the two you've mentioned are on Collectives. They should also be migrated, but at a lower priority :)

**Comment by PolkadotDom** [comment]
> This issue is focusing on the pallets being migrated from the Relay Chain to Asset Hub in the coming migration, the two you've mentioned are on Collectives. They should also be migrated, but at a lower priority :)

🙌 🙌

**Comment by PolkadotDom** [comment]
I'll take conviction voting 👍

**Comment by dharjeezy** [comment]
> I'll take conviction voting 👍

I already worked on it here initially, https://github.com/paritytech/polkadot-sdk/pull/6621

**Comment by dharjeezy** [comment]
FYI, I already worked on these pallets: [pallet conviction voting](https://github.com/paritytech/polkadot-sdk/pull/6621),
[pallet society](https://github.com/paritytech/polkadot-sdk/pull/6623), and [pallet nomination pool](https://github.com/paritytech/polkadot-sdk/pull/6715)

**Comment by muharem** [comment]
@dharjeezy @Doordashcon would you like to work on `pallet_election_provider_multi_phase` pallet?

I think I went through all open PRs, skipped only the ones that have unsolved requests for change. You can expect my reviews as soon as you solve them. It would be very helpful if we can get this issue closed before the release (mid Feb).

**Comment by dharjeezy** [comment]
> [@dharjeezy](https://github.com/dharjeezy) [@Doordashcon](https://github.com/Doordashcon) would you like to work on `pallet_election_provider_multi_phase` pallet?
> 
> I think I went through all open PRs, skipped only the ones that have unsolved requests for change. You can expect my reviews as soon as you solve them. It would be very helpful if we can get this issue closed before the release (mid Feb).

I will take pallet_election_provider_multi_phase and work on it @muharem

**Comment by muharem** [comment]
@dharjeezy @Doordashcon @seemantaggarwal please make sure you provide a proper documentation. 

Example:

```
/// Provider for the block number.
///
/// Normally this is the `frame_system` pallet.
type BlockNumberProvider: BlockNumberProvider;
```

This explanation lacks enough detail for someone encountering this feature for the first time. What is a block provider, and what is it used for? Same comment for PRdocs

The term “Normally” is also unclear - what exactly qualifies as a “normal” setup? Instead, it should explicitly state that to maintain the same behavior as before this feature was introduced, the setting should be configured as System. However, even with that clarification, users still need more context to confidently set it to System without second-guessing.

Additionally, as I mentioned in the issue:

> Update documentation for calls with block number parameters, configuration settings like BountyUpdatePeriod, and storage fields. Documentation should clarify for users that block numbers refer to those from Config::BlockNumberProvider, which may come from either the local parachain or the Relay Chain.

Right now, clients cannot assume the block number is always local - it may come from another source. This crucial detail is missing, meaning client developers will only discover it after making a mistake, often after spending considerable time troubleshooting.

I won’t block these PRs since we need them for the next release, but improving the documentation is not a major effort - a single PR could address this properly for all pallets.

**Comment by Doordashcon** [comment]
@muharem How about the migrations as @bkchr suggested?

**Comment by muharem** [comment]
@dharjeezy when you can open PR for `pallet_election_provider_multi_phase`?

**Comment by dharjeezy** [comment]
> [@dharjeezy](https://github.com/dharjeezy) when you can open PR for `pallet_election_provider_multi_phase`?

i will open this by weekend.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/asset-hub-migration-2025/11129/18
